[Flaky Test]  //staging/src/k8s.io/apiserver/pkg/storage/tests/go_default_test:run_1_of_2 - TestWatchBookmarksWithCorrectResourceVersion

<!-- Please only use this template for submitting reports about flaky tests or jobs (pass or fail with no underlying change in code) in Kubernetes CI -->

**Which jobs are flaking**:
bazel-test-master

**Which test(s) are flaking**:
//staging/src/k8s.io/apiserver/pkg/storage/tests/go_default_test:run_1_of_2 - TestWatchBookmarksWithCorrectResourceVersion

**Testgrid link**:
https://testgrid.k8s.io/sig-release-master-blocking#bazel-test-master

**Reason for failure**:

https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/periodic-kubernetes-bazel-test-master/1309347113775140864

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //staging/src/k8s.io/apiserver/pkg/storage/tests:go_default_test
-----------------------------------------------------------------------------
WARNING: Package "github.com/golang/protobuf/protoc-gen-go/generator" is deprecated.
	A future release of golang/protobuf will delete this package,
	which has long been excluded from the compatibility promise.

WARNING: Package "github.com/golang/protobuf/protoc-gen-go/generator" is deprecated.
	A future release of golang/protobuf will delete this package,
	which has long been excluded from the compatibility promise.

I0925 04:35:29.898635  111082 client.go:360] parsed scheme: "endpoint"
I0925 04:35:29.898755  111082 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:15919796450108341060  <nil> 0 <nil>}]
I0925 04:35:30.073140  111082 client.go:360] parsed scheme: "endpoint"
I0925 04:35:30.073260  111082 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:10838886466973350390  <nil> 0 <nil>}]
I0925 04:35:30.244068  111082 client.go:360] parsed scheme: "endpoint"
I0925 04:35:30.244236  111082 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:34955294453370098510  <nil> 0 <nil>}]
I0925 04:35:30.470616  111082 client.go:360] parsed scheme: "endpoint"
I0925 04:35:30.470835  111082 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:61748576491532147930  <nil> 0 <nil>}]
I0925 04:35:33.483970  111082 trace.go:205] Trace[1198431849]: "cacher list" type:*example.Pod (25-Sep-2020 04:35:30.483) (total time: 3000ms):
Trace[1198431849]: [3.000357395s] [3.000357395s] END
I0925 04:35:33.612127  111082 client.go:360] parsed scheme: "endpoint"
I0925 04:35:33.612254  111082 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:72897887250632030410  <nil> 0 <nil>}]
E0925 04:35:33.615636  111082 cacher.go:416] cacher (*example.Pod): unexpected ListAndWatch error: failed to list *example.Pod: injected error; reinitializing...
I0925 04:35:35.825090  111082 client.go:360] parsed scheme: "endpoint"
I0925 04:35:35.825225  111082 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:49485330995282495840  <nil> 0 <nil>}]
I0925 04:35:36.149345  111082 client.go:360] parsed scheme: "endpoint"
I0925 04:35:36.149471  111082 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:31634362762982687630  <nil> 0 <nil>}]
I0925 04:35:36.305832  111082 client.go:360] parsed scheme: "endpoint"
I0925 04:35:36.305957  111082 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:2607704431731553580  <nil> 0 <nil>}]
I0925 04:35:36.485113  111082 client.go:360] parsed scheme: "endpoint"
I0925 04:35:36.485260  111082 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:54730643923908686610  <nil> 0 <nil>}]
I0925 04:35:42.686841  111082 client.go:360] parsed scheme: "endpoint"
I0925 04:35:42.686992  111082 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:59629701032712351470  <nil> 0 <nil>}]
I0925 04:35:43.163978  111082 client.go:360] parsed scheme: "endpoint"
I0925 04:35:43.164125  111082 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:64660313369909079820  <nil> 0 <nil>}]
I0925 04:35:43.366229  111082 client.go:360] parsed scheme: "endpoint"
I0925 04:35:43.366366  111082 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:35564721713282456450  <nil> 0 <nil>}]
W0925 04:35:43.413038  111082 feature_gate.go:235] Setting GA feature gate WatchBookmark=true. It will be removed in a future release.
I0925 04:35:43.563189  111082 client.go:360] parsed scheme: "endpoint"
I0925 04:35:43.563317  111082 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:75027661759177906480  <nil> 0 <nil>}]
W0925 04:35:47.104324  111082 feature_gate.go:235] Setting GA feature gate WatchBookmark=true. It will be removed in a future release.
W0925 04:35:47.104866  111082 feature_gate.go:235] Setting GA feature gate WatchBookmark=true. It will be removed in a future release.
I0925 04:35:51.173891  111082 client.go:360] parsed scheme: "endpoint"
I0925 04:35:51.174037  111082 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:15536036555139789380  <nil> 0 <nil>}]
{"level":"warn","ts":"2020-09-25T04:35:54.178Z","caller":"clientv3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"endpoint://client-c161e732-de44-40a9-917d-3286765864f5/localhost:15536036555139789380","attempt":0,"error":"rpc error: code = Unavailable desc = transport is closing"}
W0925 04:35:54.178828  111082 server.go:1313] grpc: Server.processUnaryRPC failed to write status: connection error: desc = "transport is closing"
W0925 04:35:54.296890  111082 feature_gate.go:235] Setting GA feature gate WatchBookmark=true. It will be removed in a future release.
==================
WARNING: DATA RACE
Write at 0x00c000359102 by goroutine 156:
  testing.tRunner()
      GOROOT/src/testing/testing.go:1130 +0x21a

Previous write at 0x00c000359102 by goroutine 139:
  testing.(*common).FailNow()
      GOROOT/src/testing/testing.go:723 +0x4f
  testing.(*common).Fatalf()
      GOROOT/src/testing/testing.go:798 +0x9d
  staging/src/k8s.io/apiserver/pkg/storage/tests/go_default_test.TestWatchBookmarksWithCorrectResourceVersion.func1()
      staging/src/k8s.io/apiserver/pkg/storage/tests/cacher_test.go:951 +0x4a4

Goroutine 156 (running) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:1178 +0x796
  testing.runTests.func1()
      GOROOT/src/testing/testing.go:1449 +0xa6
  testing.tRunner()
      GOROOT/src/testing/testing.go:1127 +0x202
  testing.runTests()
      GOROOT/src/testing/testing.go:1447 +0x5aa
  testing.(*M).Run()
      GOROOT/src/testing/testing.go:1357 +0x4eb
  main.main()
      bazel-out/k8-fastbuild/bin/staging/src/k8s.io/apiserver/pkg/storage/tests/go_default_test_/testmain.go:120 +0x304

Goroutine 139 (finished) created at:
  staging/src/k8s.io/apiserver/pkg/storage/tests/go_default_test.TestWatchBookmarksWithCorrectResourceVersion()
      staging/src/k8s.io/apiserver/pkg/storage/tests/cacher_test.go:942 +0x664
  testing.tRunner()
      GOROOT/src/testing/testing.go:1127 +0x202
==================
--- FAIL: TestWatchBookmarksWithCorrectResourceVersion (7.19s)
    cacher_test.go:951: failed to create pod foo-62: rpc error: code = Unavailable desc = transport is closing
    testing.go:1042: race detected during execution of test
FAIL
```


**Anything else we need to know**:
- links to go.k8s.io/triage appreciated
Somehow nothing seen on triage - wonder why?

- links to specific failures in spyglass appreciated

<!-- Please see the deflaking doc (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/flaky-tests.md) for more guidance! -->
/sig testing

--- a/staging/src/k8s.io/apiserver/pkg/storage/tests/cacher_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/tests/cacher_test.go
@@ -22,6 +22,7 @@ import (
 	"reflect"
 	goruntime "runtime"
 	"strconv"
+	"sync"
 	"testing"
 	"time"
 
@@ -941,8 +942,12 @@ func TestWatchBookmarksWithCorrectResourceVersion(t *testing.T) {
 	defer watcher.Stop()
 
 	done := make(chan struct{})
-	defer close(done)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer wg.Wait()   // We must wait for the waitgroup to exit before we terminate the cache or the server in prior defers
+	defer close(done) // call close first, so the goroutine knows to exit
 	go func() {
+		defer wg.Done()
 		for i := 0; i < 100; i++ {
 			select {
 			case <-done:

